---
title: "PROJECT: CANCER - OS?DFS Analysis by Gene Using TCGA RNASeqV2 RSEM Data on TISSUE TYPE HERE"
author: "Amy Olex"
date: "DATE"
output: 
  html_document:
    toc: yes
    toc_depth: 3
---

```{r global_vars, echo=F, error=F, warning=F, message=F}
library(knitr) 
#opts_chunk$set(cache.path='cache/', fig.path='img/', cache=F, tidy=T, fig.keep='high', echo=F, error=F, warning=F, message=F, dpi=300, out.width=700)
opts_chunk$set(echo=F, error=F, warning=F, message=F)
#options(replace.assign=TRUE, width=120)
#set.seed(1)
```

```{r, echo=TRUE}
CANCER <- "PAAD"
TISSUE_CODE <- "01"  ### 01 for Primary tumor, 11 for solid normal tissue...see line 123 below.  This limits this analysis to one tissue type only.
SURVIVAL <- "OS"
CODE <- c("LIVING"=0, "DECEASED"=1)  ##categories from cBioPortal
PVAL <- .05
GROUP_MIN <- 20
STEP_SIZE <- .1 #set to zero to use NUM_STEPS
NUM_STEPS <- 0  #set to zero to use STEP_SIZE
CENSOR <- 0  #set to 0 if you don't want to censor data at a certain time point

ANALYSIS_ID <- "tumor"  ###or normal or censored etc.
WORKING_DIR <- "~/Desktop/CCTR/Data/KazuakiTakabe_S1P_0714_TCGA/TCGA Analyses/PAAD_Survival"

EXP_FILE <- "~/Desktop/CCTR/Data/KazuakiTakabe_S1P_0714_TCGA/TCGA Analyses/RNAseq_PAAD_all_111914/normalization/paad_111914_tumorOnlySamplesNormalized_n179.RData"

if(STEP_SIZE > 0){
  OUTPUT_FILE <- paste(Sys.Date(), "_survival_pvalTables_stepSize", as.character(STEP_SIZE), "_", ANALYSIS_ID, "_",CANCER,"_",SURVIVAL,".txt", sep="")
}
else{
  OUTPUT_FILE <- paste(Sys.Date(), "_survival_pvalTables_numSteps", as.character(NUM_STEPS), "_", ANALYSIS_ID, "_",CANCER,"_",SURVIVAL,".txt", sep="")
}



alias = data.frame(row.names=c( "EMR1.2015","TNF.7124","IL6.3569","STAT3.6774","NFKB1.4790","NFKB2.4791","RELA.5970","PECAM1.5175","LYVE1.10894","SPHK1.8877","SPHK2.56848","S1PR1.1901","S1PR2.9294","S1PR3.1903","S1PR4.8698","S1PR5.53637","ABCC1.4363","ABCG2.9429","SPNS2.124976","FLT1.2321","KDR.3791","FLT4.2324","TIE1.7075","TEK.7010","VEGFA.7422","VEGFC.7424","ANGPT2.285","SGPL1.8879","SPP1.6696", "ABCC11.85320"), alias=c("F4/80","TNF-alpha","IL-6","Stat3","NFkB1","NFkB2","p65","CD31","Lyve-1","SphK1","SphK2","S1PR1","S1PR2","S1PR3","S1PR4","S1PR5","ABCC1","ABCG2","Spns2","VEGFR1","VEGFR2","VEGFR3","Tie1","Tie2","VEGFA","VEGFC","Ang2","S1P-Lyase","SPP1", "ABCC11"))
```

```{r observations}
OBSERVATIONS <- "To fill in...."

```

```{r setup}
# Load needed libraries
library(plyr)
library(survival)
library(rms)

if(SURVIVAL=="OS"){
  LOWCOLOR <- "darkorchid4"
  HIGHCOLOR <- "chartreuse4"
} else {
  LOWCOLOR <- "blue"
  HIGHCOLOR <- "red"
}

# Load Survival data from file and add a type column
setwd(WORKING_DIR)

## Load the gene expression and survival data for samples (survival data is all in the same opbject now)
## Note: I may need to recode the names for what I use in this report.
load(EXP_FILE)

#coldata <- coldata1_TP             ##If recode is needed
#qnorm_data <- qnorm_data_TP        ##If recode is needed

#get the survival data
tcga <- coldata[,c("OS.status","OS.time")]  ##These headers should be generic, but if not I need to select for them specifically here.
names(tcga) <- c("event","time")
tcga$type = as.numeric(as.character(revalue(tcga$event, CODE)))
tcga$time = as.numeric(as.character(tcga$time))

# Censor data that is >CENSOR to be names(CODE[1]) with a type of CODE[[1]] and a time of CENSOR
if(CENSOR>0){
  censor <- which(tcga$time>CENSOR)
  tcga$event[censor] <- names(CODE[1])
  tcga$time[censor] <- CENSOR
  tcga$type[censor] <- CODE[[1]]
}

```

```{r functions}
makeTransparent<-function(someColor, alpha=100)
{
    newColor<-col2rgb(someColor)
    apply(newColor, 2, function(curcoldata){rgb(red=curcoldata[1], green=curcoldata[2],
                                                blue=curcoldata[3],alpha=alpha, maxColorValue=255)})
}
```

```{r specify_targets}
# transpose the vsd_data matrix
ndata_t = as.data.frame(t(qnorm_data))

#rename the genes to remove the pipe | in the name for easier parsing.
newnames = gsub("\\|", ".", colnames(ndata_t))
colnames(ndata_t) = newnames

#Extract out the genes of interest from vsd_data_t and transpose it again at the same time so the genes are in the rows again
ndata_gene = as.data.frame(t(ndata_t[,row.names(alias)]))

#Now merge this with the Alias object and get the alias names as the row names
ndata_gene2 = merge(alias, ndata_gene, by=0, all=TRUE)
row.names(ndata_gene2) = ndata_gene2$alias
ndata_gene2 = ndata_gene2[,-(1:2)]

## Transpose the matrix so genes are now the columns
gene_exp <- as.data.frame(t(ndata_gene2))

## Remove any rows that are not solid "-01" tumors because they will introduce duplicate row names.
## First identify any entries that are not solid tumors with a -01 at end.
tmp = substr(row.names(gene_exp), start=14, stop=15)
cropped_gene_exp = gene_exp[which(tmp==TISSUE_CODE),]  ###This limits this analysis to one tissue code only

## Re calculate the means
means <- lapply(cropped_gene_exp, mean)

## Crop the row names so I can merge data frames
row.names(cropped_gene_exp) = substr(row.names(cropped_gene_exp), start=1, stop=12)

## Create the high/low discretization using the mean expression for a gene across all tumor samples 
high_low <- as.data.frame(ifelse(cropped_gene_exp <= means, "low", "high"))

## Merge the high/low values with the survival data, and only keep those that are in both sets 
tcga_surv_genes = merge(tcga, high_low, by=0, all=FALSE)
row.names(tcga_surv_genes) <- tcga_surv_genes$Row.names
tcga_surv_genes <- tcga_surv_genes[,-1]

N <- dim(tcga_surv_genes)[1]

# Create the Surv() object using time and type, which is really alive or dead status.
tcga_surv <- Surv(as.numeric(paste(tcga$time)), as.numeric(paste(tcga$type)))

### Now we are ready to plot stuff
```
#Overview

This report contains the `r SURVIVAL` Kaplan-Meier survival plots for the `r CANCER` TCGA data that have both clinical and RNASeqV2 RSEM data available.    

**Additional Associated Files**

The following file contains supplementary data from this report in a tab-delimited format.  

`r OUTPUT_FILE`

**File Description:** To identify the best gene expression cutoff for each gene we tested a range of values between the minimum and maximum expression.  This file contains all gene expression cutoffs tested as well as the corresponding Log Rank P-value from the survival analysis.  The chosen cut off for each gene was that which obtained the lowest log-rank p-value regardless of statistical significance.

**Note:** The Log Rank p-values have not been corrected for multiple testing correction because their significance was not needed to obtain a cutoff for the double gene survival analysis.

##Obtaining Data

Level 3 RSEM RNASeqV2 and clinical data for all `r CANCER` patients were downloaded in bulk from Broad Institutes GDAC FireBrowse Spring 2016 Analysis Run (http://firebrowse.org) on 6/20/16.  Data was loaded into an in-house R program for analysis and pre-processing (see methods below).  The `r SURVIVAL` survival data was downloaded from cBioPortal on 6/20/16 via their DATA SETS analysis tool.

**FireBrowse References**
Bulk downloades obtained from FireBrowse Spring 2016 Analysis Run (http://firebrowse.org)

**cBioPortal References**
Cerami et al. The cBio Cancer Genomics Portal: An Open Platform for Exploring Multidimensional Cancer Genomics Data. Cancer Discovery. May 2012 2; 401. 

Gao et al. Integrative analysis of complex cancer genomics and clinical profiles using the cBioPortal. Sci. Signal. 6, pl1 (2013).

**TCGA Acknowledgement**
The results shown here are in whole or part based upon data generated by the TCGA Research Network: http://cancergenome.nih.gov/.

**R Reference**
R Core Team: R: A Language and Environment for Statistical Computing, Https://Www.R-Project.Org/. Vienna, Austria, R Foundation for Statistical Computing, 2015

##Patient Overview

There are `r dim(cropped_gene_exp)[1]` `r CANCER` patients with primary tumor RNASeqV2 data, and `r dim(tcga)[1]` patients with usable `r SURVIVAL` time points. Intersecting these data leaves `r dim(tcga_surv_genes)[1]` patients to generate the survival curves.

##Methods

**Preprocessing and Normalization**

The RSEM-derived TPM (transcript-per-million) values were calculated as described in the RSEM manuscript [see RSEM Ref].  The file "illuminahiseq_rnaseqv2-RSEM_genes" was downloaded from FireBrowse as previously described and the “scaled_estimate” data was extracted and converted to the TPM value by multiplying by 1E6.  

The TPM values were then log2 transformed ( log2( TPM + 1 ) ) before being normalized across all samples and tissue types using simple quantile normalization with the normalizeQuantiles() function in the limma R package (see Limma Ref).  This normalization was done to ensure all column sums were equivalent.  When calculating a traditional TPM value all column sums should be equivalent; however, the RSEM-derived TPM values do not have this property, so normalization was used to make the samples comparable.  Using the TPM values also allows one to compare genes with each other as it already incorporates a correction for differing gene lengths.

**Survival Analysis**

The survival analysis was performed using the R survival package (see survival Ref).  Survival plots were generated for all patients together as well as for groups of patients based on the gene expression of each gene of interest. I labeled the gene expression of each gene for each patient as either high or low with respect to a chosen cutoff.  Cutoffs were chosen in 2 ways:

- *Range of Values:* Cutoffs were chosen in increments of .1 from the minimum to the maximum gene expression of a given gene. Genes were labeled as high if the expression was greater than the cutoff and low otherwise.  If the survival pvalue for a cutoff value was <= `r PVAL` AND both highN and lowN were greater than or equal to `r GROUP_MIN` the survival plot was created, otherwise, no plot is shown.  For plots not shown you can find the Log Rank Pvalue results from the survival analysis in the output file `r OUTPUT_FILE`.  **Note** that this method will test a different number of cutoffs for each gene depending on the gene's expression range.  Genes with a larger range will have more cutoffs tested, while genes with a smaller range will have fewer cutoffs tested.
- *Mean Expression:* For each gene of interest I took the average gene expression over all tumor samples and used this as the cutoff for that gene.  All plots with survival pvalues, regardless of significance, are shown. The means were calculated over all available normal samples, but the plots only contain data for the intersecting patients.

*Censoring*

If `r CENSOR` is greater than 0 then this report contains artificially censored survivial data at `r CENSOR` days.  Any value greater than `r CENSOR` days is changed to "`r names(CODE[1])`" and with an event time of `r CENSOR` days.  

The following are the number of patients that were censored at `r CENSOR` days:
```{r}
if(CENSOR>0) print(length(censor))else print(0)
```

**References**

*RSEM*
RSEM: accurate transcript quantification from RNA-Seq data with or without a reference genome. BMC bioinformatics, Vol. 12, No. 1. (04 August 2011), 323, doi:10.1186/1471-2105-12-323 by Bo Li, Colin N. Dewey

*R Limma*
Ritchie ME, Phipson B, Wu D, et al: Limma powers differential expression analyses for RNA-sequencing and microarray studies. Nucleic Acids Res 43:e47, 2015

*R survival*
Therneau T: A package for survival analysis in S, v2.38, April 27th, 2016 update. http://CRAN.R-project.org/package=survival


#Results

##`r SURVIVAL` Curve (no division by gene expression)

```{r DFScurve}
tcga_fit <- survfit(tcga_surv~1)
plot(tcga_fit, main=paste(CANCER, SURVIVAL, "Curve, N=", N), xlab="Days to Event", ylab="Cumulative Survival Probability")
```

##Genes of Interest Survival Curves and Expression Distributions

The following graphs were made by testing a range of cutoffs for each gene.  Cutoff values range from the min to the max gene expression values for a gene in increments of 0.1.  The plots shown are those that had at least `r GROUP_MIN` patients in each high/low group AND had the most significant p-value that was at least less than the unadjusted p-value of `r PVAL`.  If a gene did not have a significant p-value then a plot is not shown.

**Tables of all pvalues for each gene and cutoff are provided in a seperate text file named `r OUTPUT_FILE` that can be opened in Excel.**


```{r incrementalPlots, fig.width=15, fig.height=5}

## Get the transpose of the data for the density plots
vsd_t <- as.data.frame(t(ndata_gene2))

## make a KM Plot for each Gene of Interest and calculate statistics
for(gene in names(cropped_gene_exp)){
  par(mfrow=c(1,2))
  # Find the min and max values for this gene
  # create a vector of values to loop over
  # Loop over values
  #     create high_low matrix
  #     merge high/low with tcga_surv_genes
  #     create survival object
  #     calculate pvalues and test for significance
  #     if significant, plot survival and distribution with data line
  #     if not significant move on to next value
  
  # create vector of X values between min and max using either NUM_STEPS or STEP_SIZE
  if(STEP_SIZE > 0){
    vals <- seq(range(cropped_gene_exp[gene])[1],range(cropped_gene_exp[gene])[2], by=STEP_SIZE)
    }
  else if(NUM_STEPS > 0){
    steps <- (range(cropped_gene_exp[gene])[2]-range(cropped_gene_exp[gene])[1])/(NUM_STEPS+1)
    vals <- seq(range(cropped_gene_exp[gene])[1],range(cropped_gene_exp[gene])[2], by=steps)[-1]
    vals <- vals[-length(vals)]
  }
  
  best <- NULL

  # Loop over vals
  for(v in vals){

    HL <- as.data.frame(ifelse(cropped_gene_exp[gene] <= v, "low", "high"))
    
    ## Merge the high/low values with the survival data, and only keep those that are in both sets
    tcga_surv_genes2 = merge(tcga, HL, by=0, all=FALSE)
    row.names(tcga_surv_genes2) <- tcga_surv_genes2$Row.names
    tcga_surv_genes2 <- tcga_surv_genes2[,-1]
    
    # Test to see if we have multiple groups
    if((length(count(tcga_surv_genes2[gene][[1]])$freq)<=1)){
      next
    }
    ## Create the Surv() object using time and type, which is really alive or dead status.
    tcga_surv <- Surv(as.numeric(paste(tcga_surv_genes2$time)), as.numeric(paste(tcga_surv_genes2$type)))
    
    ## Calculate the pvalue and test for significance at p=0.05
    sdf <- survdiff(tcga_surv~tcga_surv_genes2[gene][[1]])
    #p.val_adjust <- p.adjust( (1 - pchisq(sdf$chisq, length(sdf$n) - 1)), method=PVAL_ADJUST, n=dim(alias)[1] )
    p.val <- (1 - pchisq(sdf$chisq, length(sdf$n) - 1))
    highN <- count(tcga_surv_genes2[,gene])$freq[1]
    lowN <- count(tcga_surv_genes2[,gene])$freq[2]
    #plot both KM and density plots
    tcga_fit = survfit(tcga_surv~tcga_surv_genes2[gene][[1]])
    tcga_npfit = npsurv(tcga_surv~tcga_surv_genes2[gene][[1]])
    if((p.val <= PVAL) && (highN>=GROUP_MIN) && (lowN>=GROUP_MIN)  ){
      
      if(is.null(best)){
        #best <- list(gene=gene, fitobj=tcga_fit, cutoff=v, pval=p.val, high=highN, low=lowN, adjusted=p.val_adjust)
        best <- list(gene=gene, fitobj=tcga_npfit, cutoff=v, pval=p.val, high=highN, low=lowN)
      }
      else if(best$pval > p.val){
        #best <- list(gene=gene, fitobj=tcga_fit, cutoff=v, pval=p.val, high=highN, low=lowN, adjusted=p.val_adjust)
        best <- list(gene=gene, fitobj=tcga_npfit, cutoff=v, pval=p.val, high=highN, low=lowN)
      }
    }
  } #end for v in vals      
  
  if(!is.null(best)){
      #plot(best$fitobj, conf.int=TRUE, col=c("red", "blue"), main=paste(SURVIVAL, "(N=",N,", highN=",best$high,", lowN=",best$low,") -- ", gene, "\nCutoff=",round(best$cutoff, 4),", Unadjusted Pval = ", round(best$pval, 5),"\n", PVAL_ADJUST, " Adjusted Pval = ", round(best$adjusted, 5)), xlab="days to event", ylab="% Patients")
      #legend("topright", legend=c("High", "Low"), col=c("red", "blue"), lty=1)
  
      survplot(fit=best$fitobj,type = "kaplan-meier",conf = "bands",col.fill=c(makeTransparent(HIGHCOLOR, alpha=30), makeTransparent(LOWCOLOR, alpha=30)),label.curves = TRUE,levels.only = TRUE,time.inc = 250,dots = FALSE, n.risk = TRUE, sep.n.risk=.055, cex.n.risk=.8, adj.n.risk=.5, col=c(HIGHCOLOR, LOWCOLOR), lty=1, lwd=2, xlab="Months to Event", ylab="% Patients Still at Risk")
      
      title(paste(SURVIVAL, "(N=",N,", highN=",best$high,", lowN=",best$low,") -- ", gene, "\nCutoff=",round(best$cutoff, 4),", Log Rank Pval = ", round(best$pval, 5)))
      
      legend(x=1, y=.35, legend=c("High", "Low"), col=c(HIGHCOLOR, LOWCOLOR), lty=1, lwd=3, seg.len=1, cex=.7)

      plot(density(as.numeric(vsd_t[gene][[1]])), main=paste("Gene Expression Distribution -- ", gene))
      abline(v=mean(as.numeric(vsd_t[gene][[1]])), col="blue")
      abline(v=median(as.numeric(vsd_t[gene][[1]])), col="red")
      abline(v=best$cutoff, col="green")
      legend("topright", legend=c("Mean", "Median", "Cutoff"), col=c("blue", "red", "green"), lty=1)
   }   

} #end for each gene

```

###`r SURVIVAL` Curves Using Means

The following graphs were made by using the mean gene expression across all RNAseq samples as the threshold for high vs low expression.

```{r meanPlots, fig.width=15, fig.height=5}
## make a KM Plot for each Gene of Interest and calculate statistics
for(gene in names(cropped_gene_exp)){
      par(mfrow=c(1,2))
      #plot both KM and density plots
      tcga_fit = survfit(tcga_surv~tcga_surv_genes[gene][[1]])
      sdf <- survdiff(tcga_surv~tcga_surv_genes[gene][[1]])
      #p.val_adjust <- p.adjust( (1 - pchisq(sdf$chisq, length(sdf$n) - 1)), method=PVAL_ADJUST, n=dim(alias)[1] )
      p.val <- (1 - pchisq(sdf$chisq, length(sdf$n) - 1))
      ## Get the N for high and low
      highN <- count(tcga_surv_genes[gene])$freq[1]
      lowN <- count(tcga_surv_genes[gene])$freq[2]
      plot(tcga_fit, conf.int=TRUE, col=c("red", "blue"), main=paste(SURVIVAL, " (N=",N,", highN=",highN,", lowN=",lowN,") -- ", gene, "\nLog Rank Pval = ", round(p.val, 5)), xlab="Months to Event", ylab="% Patients Still at Risk")
      legend("topright", legend=c("High", "Low"), col=c("red", "blue"), lty=1)
  
      plot(density(as.numeric(vsd_t[gene][[1]])), main=paste("Gene Expression Distribution -- ", gene))
      abline(v=mean(as.numeric(vsd_t[gene][[1]])), col="blue")
      abline(v=median(as.numeric(vsd_t[gene][[1]])), col="red")
      legend("topright", legend=c("Mean", "Median"), col=c("blue", "red"), lty=1)

}
```

```{r, create_pval_table}
source("createPvalTable.R")
createPvalFile(TCGA_SURV_DATA=tcga, EXP_DATA=qnorm_data, CODE=CODE, alias=alias, STEP_SIZE=STEP_SIZE, FILE_NAME=OUTPUT_FILE, TISSUE_CODE)
```

